timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/ArrayLockEventSlidingWindow.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock.watch;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import com.google.common.math.LongMath;
+import com.google.common.primitives.Ints;
+import com.palantir.lock.watch.LockWatchEvent;
+
+public class ArrayLockEventSlidingWindow {
```

Comment:
The concurrency model of this class needs a write-up / comments.

It looks like `add` performs a volatile write to `nextSequence` while `getFromVersion` performs a volatile read. Then, we do check anything we return. So _I think_ this is fine, but let's talk through.

Reply:
Specifically, the cases I was concerned about involve liveness (e.g. if nextSequence was non-volatile it is possible to get into a state where the sliding window rejects all queries).

Reply:
I've added some documentation about it. PLMK if it can be improved!